Support WooCommerce-Wear in `fastlane` & Buildkite configuration

wzieba opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds Fastlane & Buildkite configuration to support releasing the WooCommerce Wear OS application.

### Testing instructions
<!-- Step-by-step testing instructions. When necessary, break out individual scenarios that need testing, and consider including a checklist for the reviewer to go through. -->


- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Add `app` parameter to `build_bundle` lane

wzieba commented:
hi @AliSoftware 👋 can I have an early feedback? Is that what we discussed internally?

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -1206,8 +1210,9 @@ platform :android do
-    release_assets = [universal_apk_path, aab_file_path].compact
+    mobile_aab_file_path = File.join(PROJECT_ROOT_FOLDER, 'artifacts', aab_file_name('WooCommerce', version))
+    wear_aab_file_path = File.join(PROJECT_ROOT_FOLDER, 'artifacts', aab_file_name('WooCommerce-Wear', version))
+    release_assets = [universal_apk_path, mobile_aab_file_path, wear_aab_file_path].compact
```
Don't we have an universal apk for the wearOS app like we do for the phone app?

wzieba replied:
Right - looking at Pocket Casts, `.apk` for `watch` is available in release resources

wzieba replied:
I'm not sure how to address this at this moment - Pocket Casts adds `.apk` as release artifact but I _think_ it comes from the build and not from Google Play (like here).

AliSoftware replied:
Well as you can see from line 1206 above, we indeed have an action ~in our `release-toolkit`~ from fastlane named `download_universal_apk_from_google_play`, that we call here to download the APK of the Phone app.

I'm not sure if just changing the `APP_PACKAGE_NAME` from the one from the phone app to the one of the wearOS app (are they using different package names? I'm guessing so?) and the `version_code` from the one from Phone to the one for wear (with the offset accounted for)… is sufficient to make the same call to `download_universal_apk_from_google_play` be able to also download the universal APK of the wearOS app, but I'm hoping so(†)… and in such case I'd suggest you to add that call in this part of the `Fastfile` (to include it in the `release_assets`) as part of this PR.

If it's not enough and we'll need to make changes to our `download_universal_apk_from_google_play` action (e.g. to implement some additional parameter like `track` that the Google API needs us to provide explicitly…), then probably better to handle that in a subsequent PR (though maybe leave a `# TODO` in the `Fastfile` about it in the meantime) after we'd have had time to work on adding that support to `download_universal_apk_from_google_play` in `release-toolkit` first.

(†) To test if `download_universal_apk_from_google_play` would work with a wearOS app, you can either build a dummy WC wearOS `.aab` (with dummy versionCode, it's just for test) and upload it to GPC manually— so it appears in the "App Bundle Explorer" there — and then call that action in a test lane with the right parameters to see if it works… (and then delete the dummy `.aab` from Bundle Explorer in GPC to clean up after yourself). Or you can do the same kind of test, but on one of the existing versionCode of the PCAndroid wearOS app that already exists (instead of doing a dummy WC wear build). As long as the action ends up successfully downloading the right `.apk` in the end, that would be enough to prove that the action works even for wearOS APKs.

AliSoftware replied:
I've just tested the `download_universal_apk_from_google_play` action for you to download the APK of a wear app from GPC (the PCAndroid one) and can confirm it works out of the box 👍 

<img width="1521" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/216089/863aac37-0284-41bc-b97e-26d06668cc44">

So all you have to do is copy/paste the call to `download_universal_apk_from_google_play` that is done on line 1206 for the phone app to replicate the same for the wear app (with adjusted `version_code`) and you'll have the universal APK available on the build machine after that ready to be added to the `release_assets`.

---

One thing to note though is that the action will fail if there's no such versionCode in GPC. So if we end up deciding that not every single release of WooCommerce will contain both the phone and the wear app shipped together every time, we will have to account for the case where there might not be a wear app's `.aab` that has been uploaded to GPC at that time.
This means that we'll likely need to **only** call the `download_universal_apk_from_google_play` for the wearOS universal APK **if** we know that we have done a wearOS release build this time around. Or wrap the call to `download_universal_apk_from_google_play` in a `begin … rescue … end` Ruby block to catch any exception/failure from that call when the wearOS `.aab` (and thus universal `.apk`) is not available this time around.

All that will depend on how you plan to add support for the wear app in Fastlane after your changes here I guess. I'm not entirely clear if your current plan is to have the in particular this `create_gh_release` lane create a single release that contains the `.aab`/`.apk` of both the wear and the phone app, or if the plan is to create one separate GitHub Release for each of those (one having the `.aab`+`.apk` for the phone app, another having the `.aab`+`apk` of the wear app)? If the former, then just checking if the wear app is available (or trying to download the universal APK of the wear app but catch the exception if it fails because it's missing) would be the solution. If the latter, then I'm guessing you'll expect that `create_gh_release` private lane to also take an `app:` parameter like the other lanes, to tell it which app you want to create the GH Release for, and only call it with the `app:"WooCommerce-Wear"` parameter if there was actually a wearOS app being built this time around (and not call `create_gh_release(app:"WooCommerce-Wear", …)` at all if not)?

wzieba replied:
I'm not sure what's the use case for adding universal apk to our releases, but I think it's good for now to have single `.apk` for both wear and phone apps 👍 .

AliSoftware replied:
Universal APK is not a single APK encapsulating wear + mobile apps.

"Universal" in this case refers to what is presented as the universal build in Google Play for a given AAB, and uses "universal" in the meaning of it not being specialized for a particular architecture or device model.

We'll still have a _separate_ AAB, and thus a separate corresponding Universal APK, for each the wear and mobile app. And in fact, given how the GitHub releases are created via API as part of our `build_and_upload_to_google_play app:…` lane, and those will be called separately for the `app:WooCommerce` and `app:WooCommerce-Wear` apps, that means we'll have separate GitHub releases (and thus git tags, but also aab/apk assets) for Mobile and Wear bundles/releases.

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -686,9 +686,9 @@ platform :android do
+    aab_file_path = File.join(PROJECT_ROOT_FOLDER, 'artifacts', aab_file_name('WooCommerce', version))
     UI.error("Unable to find a build artifact at #{aab_file_path}") unless File.exist? aab_file_path
 
     upload_to_play_store(
```
iinm, as least from looking at [the `Fastfile` of PCAndroid](https://github.com/Automattic/pocket-casts-android/blob/main/fastlane/Fastfile#L309-L316), the name of the `track` parameter of that action will be different for phone vs wear apps, so you'll probably need to account for that too.

Though while PC's `Fastfile` have one constant for each track name, given the only difference in those between phone and wear is the `wear:` prefix, maybe helper method rather than constants would make things more compact?
```ruby
# @param [Symbol] app The type of app, `:wear` or `:phone`
# @param [String] track_name The track name, e.g. `'beta'` or `'production'`
def track(app:,track_name:)
  prefix = (app == :phone) ? '' : "#{app}:"
  "#{prefix}#{track_name}"
end
```
(I haven't checked in Fastlane's docs if there might be a valid explicit prefix for the phone app… maybe `beta` can also be passed as `phone:beta` or `main:beta` or something like that and is accepted by the Google Play API too? If so we might not really need the `prefix` to be conditional on `app == :phone` and can just use `#{prefix}:#{track_name}` unconditionally? 🤔 )

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -631,9 +631,9 @@ platform :android do
     # Create the file names
     version = version_name_current
-    build_bundle(version: version, flavor: 'Vanilla')
+    build_bundle(app: 'WooCommerce', version: version, flavor: 'Vanilla')
```
Could be worth using constants instead of hardcoding the `WooCommerce` and `WooCommerce-Wear` literal at each call site that needs it.

```ruby
WOOCOMMERCE_PHONE_APP = 'WooCommerce'
WOOCOMMERCE_WEAR_APP = 'WooCommerce-Wear'
```

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -860,9 +863,10 @@ platform :android do
-    name = aab_file_name(version)
-    aab_file = 'WooCommerce-vanilla-release.aab'
-    output_dir = 'WooCommerce/build/outputs/bundle/'
+    app = options[:app]
```
One could consider adding a small helper to fetch the value of `options[:app]` while ensuring the provided value is valid (e.g. check that it's one of `PHONE_APP` vs `WEAR_APP` and not something else; or allow to pass `:wear` as a parameter to those lanes requiring an `app` parameter as a shorthand to mean `WooCommerce-Wear`, and same for `:phone` meaning `WooCommerce`?

```ruby
def app_module_name!(app:)
  case app
  when :phone, PHONE_APP
    PHONE_APP
  when :wear, WEAR_APP
    WEAR_APP
  else
    UI.user_error!("Invalid `app:` parameter")
  end
end
```
(the `!` at the end of the method name is a convention in Ruby, where we include a `!` in the name of methods that might crash.

wzieba replied:
I see, it could be useful but I decided to reference type of app always by String (`MOBILE_APP` or `WEAR_APP`) so I'm not sure if it'll be needed in this case.

AliSoftware replied:
I'm ok with not using `:app` or `:phone` and always using the real module name as `String` 👍  But tbh the main idea behind that helper was the `UI.user_error!` to catch if you make a typo in the app parameter (especially when you might end up running the lane manually from your local terminal as rare one-offs to debug some automation for example) or (possibly more likely) if you forgot to add the `app:` parameter to your call of the lane, etc.

The main goal being to avoid a crash with obscure backtrace in the `Fastfile` that doesn't make it clear what the problem is when you provided an incorrect (or missing) `app:` parameter (e.g. failure in gradle to find the task `:WooCommerceWear:bundle` because you forgot the dash, or crash in the Ruby code down the line because `app` is `nil` but we didn't add extra checks for `nil` values in our `Fastfle` code…

Btw, the idea comes from [how we did it for WordPress, where we had to similarly add an `app:` parameter](https://github.com/wordpress-mobile/WordPress-Android/blob/0f7f65b4971e595c881daec00c0c7b96fe5fa632/fastlane/lanes/release.rb#L479-L484) (this time for WordPress vs JP rather than for Phone vs Wear, but same idea)

wzieba replied:
I see, makes sense. Thanks for adding this check!

AliSoftware reviewed (commented):
ℹ️ Haven't _tested_ the changes, so this is just a preliminary review from looking at the code alone.

wzieba committed:
Prefix track name if it's app other than mobile

wzieba committed:
Use constant when referencing Gradle project/directory

wzieba committed:
Remove WOOCOMMERCE prefix from constant

It's shorter now and we're in context of WooCommerce anyway

wzieba committed:
Replace reference by symbol, in favor of string

I'd prefer to not introduce a few ways of doing same thing (app identification), hence let's maybe use String for this task

wzieba committed:
Release beta and production of wear app

wzieba commented:
Thanks @AliSoftware ! Added some improvements to address comments 👍 also, I've prepared line to publish Wear app as well, but I haven't tested it.

AliSoftware committed:
Revert back to having each lane only build one app

Make `build_and_upload_release` and `build_and_upload_beta` take the `app:` as parameter and only build that specific app. That way we will be able to build them independently in CI, as well as only build and upload a wearOS release when we need one (which might not be the case for _every_ version of the mobile app)

AliSoftware committed:
Make `trigger_release_build` accept `app:` param

And also rename `release-builds.yml` pipeline to `release-build.yml` (singular)

Then:
 - Pass that param as `APP_TO_BUILD` env var to the triggered pipeline
 - Make use of that `$APP_TO_BUILD` var in the `release-build.yml` pipeline (`label` and `command`) so it passes the value to the `release-build.sh` script it calls
 - Make the `release-build.sh` script pass `app:"$1"` when it calls the `build_and_upload_google_play` lane
 - Make the `build_and_upload_google_play` lane propagate that `app:` parameter to the `build_and_upload_{beta,release}` lanes it calls

AliSoftware committed:
Add `app:` to `trigger_release_build` call sites

AliSoftware committed:
Improve YARD doc of some lanes

AliSoftware committed:
Make Wear app job happen in same CI build (opt-in)

Instead of the WEAR_APP being build by a separate `release-build.yml` CI build with a different `app:` parameter, we now have the `release-builds.yml` pipeline contain 2 jobs, one for the `MOBILE_APP` and one for the `WEAR_APP`, the latter being conditional on the `INCLUDE_WEAR_APP` env var being defined

dangermattic commented on fastlane/Fastfile:
```diff
@@ -178,16 +182,12 @@ platform :android do
   #####################################################################################
   # complete_code_freeze
-  # -----------------------------------------------------------------------------------
+  # 
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/woocommerce/woocommerce-android/blob/e582d30d81ec9bbd1cdea742e7260eb6a71a601b/fastlane/Fastfile#L185"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

dangermattic commented on fastlane/Fastfile:
```diff
@@ -178,16 +182,12 @@ platform :android do
-  # -----------------------------------------------------------------------------------
-  # Usage:
-  # bundle exec fastlane complete_code_freeze [skip_confirm:<skip confirm>]
+  # 
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/woocommerce/woocommerce-android/blob/e582d30d81ec9bbd1cdea742e7260eb6a71a601b/fastlane/Fastfile#L187"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

dangermattic commented on fastlane/Fastfile:
```diff
@@ -264,17 +265,13 @@ platform :android do
   #####################################################################################
   # new_beta_release
-  # -----------------------------------------------------------------------------------
+  # 
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/woocommerce/woocommerce-android/blob/e582d30d81ec9bbd1cdea742e7260eb6a71a601b/fastlane/Fastfile#L268"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

dangermattic commented on fastlane/Fastfile:
```diff
@@ -264,17 +265,13 @@ platform :android do
-  # -----------------------------------------------------------------------------------
-  # Usage:
-  # bundle exec fastlane new_beta_release [skip_confirm:<skip confirm>]
+  # 
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/woocommerce/woocommerce-android/blob/e582d30d81ec9bbd1cdea742e7260eb6a71a601b/fastlane/Fastfile#L271"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

dangermattic commented on fastlane/Fastfile:
```diff
@@ -332,15 +330,13 @@ platform :android do
   #####################################################################################
   # new_hotfix_release
-  # -----------------------------------------------------------------------------------
+  # 
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/woocommerce/woocommerce-android/blob/e582d30d81ec9bbd1cdea742e7260eb6a71a601b/fastlane/Fastfile#L333"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

dangermattic commented on fastlane/Fastfile:
```diff
@@ -332,15 +330,13 @@ platform :android do
-  # -----------------------------------------------------------------------------------
-  # Usage:
-  # bundle exec fastlane new_hotfix_release [skip_confirm:<skip confirm>] [version_name:<x.y.z>] [version_code:<nnnn>]
+  # 
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/woocommerce/woocommerce-android/blob/e582d30d81ec9bbd1cdea742e7260eb6a71a601b/fastlane/Fastfile#L335"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

dangermattic commented on fastlane/Fastfile:
```diff
@@ -394,15 +390,12 @@ platform :android do
   #####################################################################################
   # finalize_hotfix_release
-  # -----------------------------------------------------------------------------------
+  # 
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/woocommerce/woocommerce-android/blob/e582d30d81ec9bbd1cdea742e7260eb6a71a601b/fastlane/Fastfile#L393"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

dangermattic commented on fastlane/Fastfile:
```diff
@@ -394,15 +390,12 @@ platform :android do
-  # -----------------------------------------------------------------------------------
-  # Usage:
-  # bundle exec fastlane finalize_hotfix_release
+  # 
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/woocommerce/woocommerce-android/blob/e582d30d81ec9bbd1cdea742e7260eb6a71a601b/fastlane/Fastfile#L395"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

dangermattic commented on fastlane/Fastfile:
```diff
@@ -459,16 +453,12 @@ platform :android do
   #####################################################################################
   # finalize_release
-  # -----------------------------------------------------------------------------------
+  # 
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/woocommerce/woocommerce-android/blob/e582d30d81ec9bbd1cdea742e7260eb6a71a601b/fastlane/Fastfile#L456"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

dangermattic commented on fastlane/Fastfile:
```diff
@@ -459,16 +453,12 @@ platform :android do
-  # -----------------------------------------------------------------------------------
-  # Usage:
-  # bundle exec fastlane finalize_release [skip_confirm:<skip confirm>]
+  # 
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/woocommerce/woocommerce-android/blob/e582d30d81ec9bbd1cdea742e7260eb6a71a601b/fastlane/Fastfile#L458"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

dangermattic commented on fastlane/Fastfile:
```diff
@@ -572,26 +563,25 @@ platform :android do
   #####################################################################################
   # build_and_upload_google_play
-  # -----------------------------------------------------------------------------------
+  # 
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/woocommerce/woocommerce-android/blob/e582d30d81ec9bbd1cdea742e7260eb6a71a601b/fastlane/Fastfile#L566"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

dangermattic commented on fastlane/Fastfile:
```diff
@@ -572,26 +563,25 @@ platform :android do
-  # Example:
-  # bundle exec fastlane build_and_upload_google_play
-  #####################################################################################
+  # 
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/woocommerce/woocommerce-android/blob/e582d30d81ec9bbd1cdea742e7260eb6a71a601b/fastlane/Fastfile#L569"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

dangermattic commented on fastlane/Fastfile:
```diff
@@ -572,26 +563,25 @@ platform :android do
-  #####################################################################################
+  # 
+  # @param [String] app The Android app to build (MOBILE_APP aka 'WooCommerce' or WEAR_APP aka 'WooCommerce-Wear')
+  # 
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/woocommerce/woocommerce-android/blob/e582d30d81ec9bbd1cdea742e7260eb6a71a601b/fastlane/Fastfile#L571"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

AliSoftware committed:
Add the app being build in the CI logs

AliSoftware committed:
Fix rubocop violations

AliSoftware commented on .buildkite/release-builds.yml:
```diff
@@ -17,9 +17,19 @@ steps:
+      ".buildkite/commands/release-build.sh" "WooCommerce-Wear"
     priority: 1
     plugins: *common_plugins
+    if: build.env('INCLUDE_WEAR_APP') == "true"
```
See D149978-code: This env var is passed by the ReleasesV2 scenario during the calls to scenario `Tasks` which are triggering `complete_code_freese`, `new_beta_release`, `finalize_release` and `finalize_hotfix_release` (for hotfix scenarios), to decide if that CI job building the wear app will have to be included (alongside the CI job building the mobile app which is always there) in the CI build or not.

AliSoftware replied:
Actually sorry but that was a bit of a misleading comment here: the `INCLUDE_WEAR_APP` that is passed here doesn't directly come from ReleaseV2 when triggering a pipeline.

Instead, ReleasesV2 triggers the `release-pipelines/*.yml` ones, like `release-pipelines/new-beta-release.yml` for example, and it's that pipeline to which ReleasesV2 is passing the correct `INCLUDE_WEAR_APP` value to.

That environment var is then used in line 12 of those `release-pipelines/*.yml` to propagate that value via a `include_wear_app=${INCLUDE_WEAR_APP}` argument passed to the call of the `bundle exec new_beta_release …` call, which is doing version bumps and similar prep stuff before finally triggering a new CI build (`trigger_release_build`) doing the actual compilation (`build_bundle`) + upload to PlayStore. That 2nd/transitive CI build triggered from the first one is itself finally passed an `INCLUDE_WEAR_APP` env var (Line 711 of the Fastfile).

---

So in the end due to our approach with a "trampoline build" triggered by ReleaseV2 to do the prep work before it triggering another build, that variable propagation goes through a couple of hoops 😅:

```
[ReleaseV2]
 |
 | (triggers)
 V
[release-pipelines/new-beta-release.yml with INCLUDE_WEAR_APP set as env var]
 |
 | (calls)
 V
[fastlane new_beta_release include_wear_app:$INCLUDE_WEAR_APP]
 |
 | (do version bump, then calls)
 V
[trigger_release_build include_wear_app:…]
 |
 | (uses Buildkite API to trigger a new Build with 'PIPELINE'='release-builds.yml' and 'INCLUDE_WEAR_APP'=include_wear_app)
 V
[release-builds.yml triggered with INCLUDE_WEAR_APP set as env var]
 |
 | (contains 1 or 2 CI jobs depending on build.env('INCLUDE_WEAR_APP')
 V
[each job calls build_and_upload_google_play with appropriate app: parameter
```

wzieba replied:
thanks for more context @AliSoftware !

AliSoftware commented on .buildkite/release-builds.yml:
```diff
@@ -17,9 +17,19 @@ steps:
 
-  - label: "🛠 Release Build"
-    command: ".buildkite/commands/release-build.sh"
+  - label: "🛠 Release Build (Mobile App)"
+    command: |
+      ".buildkite/commands/release-build.sh" "WooCommerce"
+    priority: 1
+    plugins: *common_plugins
+    notify:
+      - slack: "#build-and-ship"
```
Food for thoughts:

💡  I'm wondering if it would be useful to add a `if: build.env('EXCLUDE_MOBILE_APP') == "true"` condition on this job (or change the env var from being `INCLUDE_WEAR_APP:true/false` to be `INCLUDE_APPS:mobile/wear/both`?) in order to have some escape hatch to manually recover for cases where we'd need to fix only the wear app.
What I mean by that is that, while in practice it's expected the mobile app to always be included, and the Release Scenario is taylored to only handle "mobile only" or "mobile + wear" cases so would never pass `EXCLUDE_MOBILE_APP=true` in practice…, for the very rare exceptions where we might need to only do a release of the wear app without the mobile app, we'd still have the door open to trigger that **manually** (even if that means triggering a build manually via the Buildkite UI, with `PIPELINE=release-builds.yml` + `EXCLUDE_MOBILE_APP=true`, instead of using the ReleasesV2 scenario for such very rare cases, given that I don't expect this to happen often in practice and would only be to handle exceptional cases or when something when wrong and we needed to take over manually to fix things… especially for the first release(s) where things might not go as expected?) 🤔 

Again, such an option would mostly only be used for debugging / recovering from situations where something when wrong and for some reason we (Apps Infra) need to only redo a release of the wear app without the mobile app manually, taking over the developers to help then solve such situations. That being said, if that's only for those seldom cases we can always trigger a build with `INCLUDE_WEAR_APP=true`, and immediately just cancel the job for the mobile app once the CI build containing both jobs has started, to only keep the wear CI job…

But who knows, maybe one day the developers will want to genuinely have a release of the wear app that would not include the mobile app for a given sprint and that would be useful to them too… (though I doubt such a scenario would happen at least at first, given we ended up agreeing that the mobile and wear app would use the same `versionName` base, so apart from the case of hotfixes that would only need a hotfix for the wear app, I don't see a version `19.4w` of the wear app without a `19.4` of the mobile app which would imply that the mobile app would skip `19.4` and go from `19.3` to `19.5`…)

wzieba replied:
I'm not against `EXCLUDE_MOBILE_APP=true`, but I'm not sure if it's necessary. My concern is only additional complexity, for something that might be not needed - if I read correctly, Pocket Casts Android doesn't have "release wear only" option.

wzieba commented:
Both, this PR and Phabricator diffs look good to me. Great job @AliSoftware !

AliSoftware committed:
Merge branch 'feature/introduce-wear-app' into support_wear_app_fastlane
